crates/node_types/prover/src/prover/mod.rs

```diff
@@ -403,23 +439,86 @@ impl Prover {
         Ok(())
     }
 
-    async fn prove_epoch(&self, epoch_height: u64, batch: &Batch) -> Result<FinalizedEpoch> {
+    async fn prove_with_base_prover(
+        &self,
+        epoch_height: u64,
+        batch: &Batch,
+    ) -> Result<(
+        SP1ProofWithPublicValues,
+        tokio::sync::RwLockReadGuard<'_, CpuProver>,
+        &SP1VerifyingKey,
+    )> {
         let mut stdin = SP1Stdin::new();
-        stdin.write(&batch);
-        let client = self.prover_client.read().await;
+        stdin.write(batch);
+
+        let client = self.base_prover_client.read().await;
+        info!("generating proof for epoch {}", epoch_height);
+
+        #[cfg(feature = "groth16")]
+        let proof = client.prove(&self.base_proving_key, &stdin).groth16().run()?;
+        #[cfg(not(feature = "groth16"))]
+        let proof = client.prove(&self.base_proving_key, &stdin).run()?;
 
-        info!("generating proof for epoch at height {}", epoch_height);
-        let proof = client.prove(&self.proving_key, &stdin).groth16().run()?;
         info!("successfully generated proof for epoch {}", epoch_height);
+        Ok((proof, client, &self.base_verifying_key))
+    }
 
-        client.verify(&proof, &self.verifying_key)?;
+    async fn prove_with_recursive_prover(
```

Comment:
I would call these methods prove_inner and prove_outer, and add rustdoc explaining